src/rcc.rs

```diff
@@ -221,50 +235,67 @@ impl CFGR {
             })
         }
 
+        // the USB clock is only valid if an external crystal is used, the PLL is enabled, and the
+        // PLL output frequency is a supported one.
+        // usbpre == false: divide clock by 1.5, otherwise no division
+        let (usbpre, usbclk_valid) = match (self.hse, pllmul_bits, sysclk) {
+            (Some(_), Some(_), 72_000_000) => (false, true),
+            (Some(_), Some(_), 48_000_000) => (true, true),
+            _ => (true, false),
+        };
+
         let rcc = unsafe { &*RCC::ptr() };
+
+        if self.hse.is_some() {
+            // enable HSE and wait for it to be ready
+
+            rcc.cr.modify(|_, w| w.hseon().set_bit());
+
+            while rcc.cr.read().hserdy().bit_is_clear() {}
+        }
+
         if let Some(pllmul_bits) = pllmul_bits {
-            // use PLL as source
+            // enable PLL and wait for it to be ready
 
             rcc.cfgr.write(|w| unsafe { w.pllmul().bits(pllmul_bits) });
 
             rcc.cr.write(|w| w.pllon().set_bit());
 
             while rcc.cr.read().pllrdy().bit_is_clear() {}
-
-            // SW: PLL selected as system clock
-            rcc.cfgr.modify(|_, w| unsafe {
-                w.ppre2()
-                    .bits(ppre2_bits)
-                    .ppre1()
-                    .bits(ppre1_bits)
-                    .hpre()
-                    .bits(hpre_bits)
-                    .sw()
-                    .bits(0b10)
-            });
-        } else {
-            // use HSI as source
-
-            // SW: HSI selected as system clock
-            rcc.cfgr.write(|w| unsafe {
-                w.ppre2()
-                    .bits(ppre2_bits)
-                    .ppre1()
-                    .bits(ppre1_bits)
-                    .hpre()
-                    .bits(hpre_bits)
-                    .sw()
-                    .bits(0b00)
-            });
         }
 
+        // set prescalers and clock source
+        rcc.cfgr.modify(|_, w| unsafe {
+            w.ppre2()
+                .bits(ppre2_bits)
+                .ppre1()
+                .bits(ppre1_bits)
+                .hpre()
+                .bits(hpre_bits)
+                .usbpre()
```

Comment:
Seems like the `stm32f301` and the `stm32f334` don't have a USB prescalers. If you'd like, I can fix this up with some conditional configuration.

Reply:
Thank you, I was not aware of this.

Reply:
I added one more commit for feature-gating this.